c/misra/src/rules/RULE-5-1/ExternalIdentifiersNotDistinct.ql

```diff
@@ -0,0 +1,42 @@
+/**
+ * @id c/misra/external-identifiers-not-distinct
+ * @name RULE-5-1: External identifiers shall be distinct
+ * @description Using nondistinct external identifiers results in undefined behaviour.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity warning
+ * @tags external/misra/id/rule-5-1
+ *       correctness
+ *       maintainability
+ *       readability
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.Linkage
+
+class ExternalIdentifiers extends Declaration {
+  ExternalIdentifiers() { hasExternalLinkage(this) }
+}
+
+class ExternalIdentifiersLong extends ExternalIdentifiers {
+  ExternalIdentifiersLong() { this.getName().length() >= 32 }
+}
+
+predicate notSame(ExternalIdentifiers d, ExternalIdentifiers d2) {
+  not d = d2 and
+  d.getLocation().getStartLine() >= d2.getLocation().getStartLine()
+}
+
+from ExternalIdentifiers d, ExternalIdentifiers d2
+where
+  not isExcluded(d, Declarations1Package::externalIdentifiersNotDistinctQuery()) and
+  notSame(d, d2) and
+  //C99 states the first 31 characters of external identifiers are significant
+  //C90 states the first 6 characters of external identifiers are significant and case is not required to be significant
+  //C90 is not currently considered by this rule
+  if d instanceof ExternalIdentifiersLong and d2 instanceof ExternalIdentifiersLong
+  then d.getName().substring(0, 30) = d2.getName().substring(0, 30)
+  else d.getName() = d2.getName()
```

Comment:
A general problem with this type of duplicated identifier rule is that there are some circumstances where we get multiple `Declaration`s appear in the database for effectively the same conceptual declaration. Examples include:

 * Compiling the same file twice under different contexts, such as different platform targets (which can result in a different `Declaration` per context).
 * Creating a dynamically linked library and indexing that in the same database as a client of that library (which can result in a `Declaration` each for the library and the client). Note: we typically don't have this issue with statically linked libraries, because we "extract" linker calls to de-duplicate client/library declarations.

I don't think there's a perfect solution to this problem, but the ones I think are most plausible are:

 1. Exclude cases where the names are identical to focus on those cases where the 31 character limit causes confusion. Pros: applicable to both functions and variables, limited false positives. Cons: will not report genuine ODR violation cases.
 2. Restrict the scope to a single link target. i.e. only report issues where we are sure that the two definitions were linked into the same binary. Pros: should eliminate false positives. Cons: we only have this information for `Function` declarations, can exclude genuinely interesting results that occur between translation units which are not linked together.
 3. Restrict the query to only report declarations where `hasDefinition()` is `true`. Pros: will avoid the common case related to dynamic libraries. Cons: false positives still possible in the multiple-compilations-of-the-same-file case.

@knewbury01 I think 1. is probably the best approach for now, but if we adopt that approach we should file an issue to look for ODR violations under Rule 5.1 at a later point.

Reply:
for my own learning:
is "Compiling the same file twice under different contexts, such as different platform targets (which can result in a different Declaration per context)." realistic in the same compilation execution? or does this mean conditional compilation, where only some src is seen per compilation? if it is the latter, again, similar to the Macros, does the DB actually have these same conceptual declarations in it?

Reply:
I have gone with solution 1 for now and opened issue to enhance in future

Reply:
> "Compiling the same file twice under different contexts, such as different platform targets (which can result in a different Declaration per context)." realistic in the same compilation execution?

This can happen in a single database if the user runs two consecutive compilations on the same machine. For example, a 32 bit build followed by a 64 bit build. In this case I believe we will get a different `Declaration` for each context (for example, a 32-bit version of a function and a 64-bit version of the function).